// native-base/lib/module/components/primitives/Stack/Stack.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { memo, forwardRef } from 'react';
import { default as Box } from '../Box';
import { getSpacedChildren } from '../../../utils';
import { usePropsResolution } from '../../../hooks/useThemeProps';
import { useHasResponsiveProps } from '../../../hooks/useHasResponsiveProps';
import { ResponsiveQueryContext } from '../../../utils/useResponsiveQuery/ResponsiveQueryProvider';
import { useToken } from '../../../hooks';

const Stack = ({
  space,
  ...props
}, ref) => {
  const {
    children,
    direction,
    reversed,
    divider,
    size,
    ...resolvedProps
  } = usePropsResolution('Stack', { ...props,
    size: space
  }, {}, {
    resolveResponsively: []
  }); //TODO: refactor for responsive prop

  if (useHasResponsiveProps(props)) {
    return null;
  } // eslint-disable-next-line react-hooks/rules-of-hooks


  const responsiveQueryContext = React.useContext(ResponsiveQueryContext);
  const disableCSSMediaQueries = responsiveQueryContext.disableCSSMediaQueries; // eslint-disable-next-line react-hooks/rules-of-hooks

  const resolvedGap = useToken('space', size);
  return /*#__PURE__*/React.createElement(Box, _extends({
    flexDirection: direction
  }, resolvedProps, {
    ref: ref,
    style: disableCSSMediaQueries ? {} : {
      gap: resolvedGap
    }
  }), getSpacedChildren(children, size, direction === 'row' ? 'X' : 'Y', reversed ? 'reverse' : 'normal', divider));
};

export default /*#__PURE__*/memo( /*#__PURE__*/forwardRef(Stack));
